private/bufpkg/bufplugin/bufpluginconfig/config.go

```diff
@@ -0,0 +1,191 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufpluginconfig
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginref"
+	// Note that the semver package we're using conforms to the
+	// support SemVer syntax found in the go.mod file. This means
+	// that runtime dependencies will need to specify the 'v' prefix
+	// in their semantic version even if it isn't directly applicable
+	// to that runtime environment (e.g. NPM).
+	//
+	// We'll use this for now so that runtime dependencies are
+	// consistent across each runtime configuration, but we might need
+	// to change this later.
+	"golang.org/x/mod/semver"
+)
+
+func newConfig(externalConfig ExternalConfig) (*Config, error) {
+	pluginIdentity, err := bufpluginref.PluginIdentityForString(externalConfig.Name)
+	if err != nil {
+		return nil, err
+	}
+	var options map[string]string
+	if len(externalConfig.Opts) > 0 {
+		// We only want to create a non-nil map if the user
+		// actually specified any options.
+		options = make(map[string]string)
+	}
+	for _, option := range externalConfig.Opts {
+		split := strings.Split(option, "=")
+		if len(split) > 2 {
+			return nil, errors.New(`plugin options must be specified as "<key>=<value>" strings`)
+		}
+		if len(split) == 1 {
+			// Some plugins don't actually specify the '=' delimiter
+			// (e.g. protoc-gen-java's java_opt=annotate_code). To
+			// support these legacy options, we map the key to an empty
+			// string value.
+			//
+			// This means that plugin options with an explicit
+			// 'something=""' option are actually passed in as
+			// --something (omitting the explicit "" entirely).
+			//
+			// This behavior might need to change depending on if
+			// there are valid use cases here, but we eventually
+			// want to support structured options as key, value
+			// pairs so we enforce this implicit behavior for now.
+			split = append(split, "")
+		}
+		key, value := split[0], split[1]
+		if _, ok := options[key]; ok {
+			return nil, fmt.Errorf("plugin option %q was specified more than once", key)
+		}
+		options[key] = value
+	}
+	runtimeConfig, err := newRuntimeConfig(externalConfig.Runtime)
+	if err != nil {
+		return nil, err
+	}
+	return &Config{
+		Name:    pluginIdentity,
+		Options: options,
+		Runtime: runtimeConfig,
+	}, nil
+}
+
+func newRuntimeConfig(externalRuntimeConfig ExternalRuntimeConfig) (*RuntimeConfig, error) {
+	var (
+		isArchiveEmpty = externalRuntimeConfig.Archive.IsEmpty()
+		isGoEmpty      = externalRuntimeConfig.Go.IsEmpty()
+		isNPMEmpty     = externalRuntimeConfig.NPM.IsEmpty()
+	)
+	if isArchiveEmpty && isGoEmpty && isNPMEmpty {
```

Comment:
Id probably do this as:

```go
if (!isArchiveEmpty && !isGoEmpty) || (!isArchiveEmpty && !isNPMEmpty) || (!isGoEmpty && !isNPMEmpty) {
  return error
}
// nil vs error, idk if we should distinguish, right?
return &RuntimeConfig {
  NPM: newNPM,
  Go: newGo,
  Archive: newArchive,
}
```

Reply:
Or, if you want to make this scale to 10 more runtimes:

```go
emptyCount := 0
for _, isEmpty := range []bool {
  externalRuntimeConfig.Archive.IsEmpty(),
  externalRuntimeConfig.Go.IsEmpty(),
  ...
} {
  if isEmpty {
    emptyCount++
  }
}
if emptyCount > 1 {
  return error
}
```

Reply:
Or, add a `pkg/boolutil` with `Count([]bool) int`

Reply:
Noted, I'll find a better way to do it 👍

Reply:
I would do it with `UnmarshalYAML` and `UnmarshalJSON`